File: frontend/src/Services/contaAPagarService.ts
import { apiFetch } from '../Utils/apiFetch';
import { API_BASE } from '../config';

const API = `${API_BASE}/api/contaapagar`;

export interface ContaAPagarDTO {
  contaAPagarId: number;
  contaId: number;
  despesaId: number;
  nomeDespesa: string;
  numeroParcela: number;
  totalParcelas: number;
  valorParcela: number;
  valorTotal: number;
  dataVencimento: string;
  dataAbertura: string;
  dataPagamento?: string;
  pago: boolean;
}

export interface ContaAPagarCreateDTO {
  despesaId: number;
  totalParcelas: number;
  valorParcela: number;
  dataPrimeiroVencimento: string;
}

export interface ContaAPagarUpdateDTO {
  valorParcela?: number;
  dataVencimento?: string;
}

export const getAll = async (): Promise<ContaAPagarDTO[]> => {
  const r = await apiFetch(API);
  return r.json();
};

export const getById = async (id: number): Promise<ContaAPagarDTO> => {
  const r = await apiFetch(`${API}/${id}`);
  return r.json();
};

export const getByContaId = async (contaId: number): Promise<ContaAPagarDTO[]> => {
  const r = await apiFetch(`${API}/conta/${contaId}`);
  return r.json();
};

export const create = async (dto: ContaAPagarCreateDTO): Promise<void> => {
  const r = await apiFetch(API, { method: 'POST', body: JSON.stringify(dto) });
  if (!r.ok) throw new Error('Erro ao lançar conta.');
};

export const update = async (id: number, dto: ContaAPagarUpdateDTO): Promise<void> => {
  await apiFetch(`${API}/${id}`, { method: 'PUT', body: JSON.stringify(dto) });
};

export const remove = async (id: number): Promise<void> => {
  await apiFetch(`${API}/${id}`, { method: 'DELETE' });
};

export const deleteConta = async (contaId: number): Promise<void> => {
  await apiFetch(`${API}/conta/${contaId}`, { method: 'DELETE' });
};

export const baixaParcela = async (id: number): Promise<void> => {
  await apiFetch(`${API}/${id}/baixa`, { method: 'PATCH' });
};

export const baixaConta = async (contaId: number): Promise<void> => {
  await apiFetch(`${API}/conta/${contaId}/baixa`, { method: 'PATCH' });
};

File: backend/PepperImportsAPI/Application/Services/ContaAPagar/ContaAPagarService.cs
using PepperImportsAPI.Application.DTOs.ContaAPagar;
using PepperImportsAPI.Application.Interfaces;
using PepperImportsAPI.Domain.Entities;

namespace PepperImportsAPI.Application.Services
{
    public class ContaAPagarService
    {
        private readonly IContaAPagarRepository _repository;
        private readonly IDespesaRepository _despesaRepository;

        public ContaAPagarService(IContaAPagarRepository repository, IDespesaRepository despesaRepository)
        {
            _repository = repository;
            _despesaRepository = despesaRepository;
        }

        public async Task Create(ContaAPagarCreateDTO dto)
        {
            var despesa = await _despesaRepository.GetByIdAsync(dto.DespesaId);
            if (despesa == null)
                throw new Exception("Despesa não encontrada.");

            var contaId = await _repository.GetNextContaIdAsync();
            var valorTotal = dto.TotalParcelas * dto.ValorParcela;
            var dataAbertura = DateTime.UtcNow;

            var parcelas = new List<ContaAPagar>();

            for (int i = 0; i < dto.TotalParcelas; i++)
            {
                parcelas.Add(new ContaAPagar
                {
                    ContaId = contaId,
                    DespesaId = dto.DespesaId,
                    NumeroParcela = i + 1,
                    TotalParcelas = dto.TotalParcelas,
                    ValorParcela = dto.ValorParcela,
                    ValorTotal = valorTotal,
                    DataVencimento = dto.DataPrimeiroVencimento.AddDays(30 * i),
                    DataAbertura = dataAbertura,
                    Pago = false
                });
            }

            await _repository.AddRangeAsync(parcelas);
        }

        public async Task Update(int id, ContaAPagarUpdateDTO dto)
        {
            var conta = await _repository.GetByIdAsync(id);
            if (conta == null)
                throw new Exception("Conta a pagar não encontrada.");

            if (dto.ValorParcela.HasValue)
                conta.ValorParcela = dto.ValorParcela.Value;

            if (dto.DataVencimento.HasValue)
                conta.DataVencimento = dto.DataVencimento.Value;

            await _repository.UpdateAsync(conta);
        }

        public async Task BaixaParcela(int id)
        {
            var conta = await _repository.GetByIdAsync(id);
            if (conta == null)
                throw new Exception("Conta a pagar não encontrada.");

            if (conta.Pago)
                throw new Exception("Parcela já foi baixada.");

            conta.Pago = true;
            conta.DataPagamento = DateTime.UtcNow;

            await _repository.UpdateAsync(conta);
        }

        public async Task BaixaConta(int contaId)
        {
            var parcelas = await _repository.GetByContaIdAsync(contaId);
            if (!parcelas.Any())
                throw new Exception("Conta não encontrada.");

            var dataPagamento = DateTime.UtcNow;

            foreach (var parcela in parcelas.Where(p => !p.Pago))
            {
                parcela.Pago = true;
                parcela.DataPagamento = dataPagamento;
                await _repository.UpdateAsync(parcela);
            }
        }

        public async Task Delete(int id)
        {
            var conta = await _repository.GetByIdAsync(id);
            if (conta == null)
                throw new Exception("Conta a pagar não encontrada.");

            await _repository.DeleteAsync(id);
        }

        public async Task DeleteConta(int contaId)
        {
            var parcelas = await _repository.GetByContaIdAsync(contaId);
            if (!parcelas.Any())
                throw new Exception("Conta não encontrada.");

            await _repository.DeleteByContaIdAsync(contaId);
        }

        public async Task<List<ContaAPagarDTO>> GetAll()
        {
            var contas = await _repository.GetAllAsync();
            return contas.Select(MapToDTO).ToList();
        }

        public async Task<ContaAPagarDTO> GetById(int id)
        {
            var conta = await _repository.GetByIdAsync(id);
            if (conta == null)
                throw new Exception("Conta a pagar não encontrada.");

            return MapToDTO(conta);
        }

        public async Task<List<ContaAPagarDTO>> GetByContaId(int contaId)
        {
            var parcelas = await _repository.GetByContaIdAsync(contaId);
            if (!parcelas.Any())
                throw new Exception("Conta não encontrada.");

            return parcelas.Select(MapToDTO).ToList();
        }

        private static ContaAPagarDTO MapToDTO(ContaAPagar conta)
        {
            return new ContaAPagarDTO
            {
                ContaAPagarId = conta.ContaAPagarId,
                ContaId = conta.ContaId,
                DespesaId = conta.DespesaId,
                NomeDespesa = conta.Despesa?.NomeDespesa,
                NumeroParcela = conta.NumeroParcela,
                TotalParcelas = conta.TotalParcelas,
                ValorParcela = conta.ValorParcela,
                ValorTotal = conta.ValorTotal,
                DataVencimento = conta.DataVencimento,
                DataAbertura = conta.DataAbertura,
                DataPagamento = conta.DataPagamento,
                Pago = conta.Pago
            };
        }
    }
}
